torch/csrc/CudaIPCTypes.cpp

```diff
@@ -0,0 +1,156 @@
+#ifdef USE_CUDA
+#include <torch/csrc/CudaIPCTypes.h>
+#include <map>
+#include <mutex>
+
+namespace torch {
+
+void BadIPCPatternError() {
+  static bool warned = false;
+  if (not warned) {
+    LOG(WARNING)
+        << "Producer process has been terminated before all shared CUDA tensors released.";
+    warned = true;
+  }
+}
+
+struct CudaIPCGlobalEntities {
+  std::mutex mutex;
+  std::map<std::string, std::shared_ptr<CudaIPCRefCountersFile>>
+      ref_counters_files_;
+  std::shared_ptr<CudaIPCRefCountersFile> next_available_ref_counters_file_;
+  CudaIPCSentDataLimbo CudaIPCSentDataLimbo_;
+  CudaIPCGlobalEntities() : ref_counters_files_() {}
+  ~CudaIPCGlobalEntities() {
+    CudaIPCSentDataLimbo_.collect();
+    safe_clean_current_file();
+    if (next_available_ref_counters_file_ != nullptr) {
+      BadIPCPatternError();
+    }
+  };
+  void safe_clean_current_file() {
```

Comment:
One thing I am not sure about: why is it only necessary to clean the current file? If we were doing some sort of global destruction thing, I would have vaguely expected that you have to loop over all the counter files.

Reply:
I guess the answer is, you're not actually maintaining the counters files, so there's no way to "clean the old ones". But then, that leads to the next question: why is it necessary to clean the current counter file, but not the old ones?

Reply:
Rest of the files going to be collected automatically by ReturnRefCounter. It is only current file which might end in state when file is open, but there is no active counters using it. Let say capacity was 10,000 we created one counter and immediately returned it. Now there is no active counters, but next_available_ref_counters_file_ still holding file handle.